src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TLV::TLVWriter checkpoint;
+    EventLoadOutContext * loadOutContext = static_cast<EventLoadOutContext *>(apContext);
+    err                                  = EventIterator(aReader, aDepth, loadOutContext);
+    if (err == CHIP_EVENT_ID_FOUND)
+    {
+        // checkpoint the writer
+        checkpoint = loadOutContext->mWriter;
+
+        err = CopyEvent(aReader, loadOutContext->mWriter, loadOutContext);
+
+        // CHIP_NO_ERROR and CHIP_END_OF_TLV signify a
+        // successful copy.  In all other cases, roll back the
+        // writer state back to the checkpoint, i.e., the state
+        // before we began the copy operation.
+        VerifyOrExit((err == CHIP_NO_ERROR) || (err == CHIP_END_OF_TLV), loadOutContext->mWriter = checkpoint);
+
+        loadOutContext->mCurrentSystemTime.mValue = 0;
+        loadOutContext->mFirst                    = false;
+        loadOutContext->mCurrentEventNumber++;
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, PriorityLevel aPriority, EventNumber & aEventNumber)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    const bool recurse = false;
+    TLVReader reader;
+    CircularEventBufferWrapper bufWrapper;
+    EventLoadOutContext context(aWriter, aPriority, aEventNumber);
+
+    CircularEventBuffer * buf = mpEventBuffer;
+    Platform::CriticalSectionEnter();
+
+    while (!buf->IsFinalDestinationForPriority(aPriority))
+    {
+        buf = buf->GetNextCircularEventBuffer();
+    }
+
+    context.mCurrentSystemTime  = buf->mFirstEventSystemTimestamp;
+    context.mCurrentEventNumber = buf->mFirstEventNumber;
+    err                         = GetEventReader(reader, aPriority, &bufWrapper);
+    SuccessOrExit(err);
+
+    err = TLV::Utilities::Iterate(reader, CopyEventsSince, &context, recurse);
+
+exit:
+    aEventNumber = context.mCurrentEventNumber;
+    Platform::CriticalSectionExit();
+
+    return err;
+}
+
+CHIP_ERROR EventManagement::GetEventReader(TLVReader & aReader, PriorityLevel aPriority, CircularEventBufferWrapper * apBufWrapper)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    CircularEventReader reader;
+    CircularEventBuffer * buffer = GetPriorityBuffer(aPriority);
+    VerifyOrExit(buffer != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    apBufWrapper->mpCurrent = buffer;
+    reader.Init(apBufWrapper);
+    aReader.Init(reader);
+exit:
+    return err;
+}
+
+CHIP_ERROR EventManagement::FetchEventParameters(const TLVReader & aReader, size_t aDepth, void * apContext)
+{
+    CHIP_ERROR err                  = CHIP_NO_ERROR;
+    EventEnvelopeContext * envelope = static_cast<EventEnvelopeContext *>(apContext);
+    TLVReader reader;
+    uint16_t extPriority; // Note: the type here matches the type case in EventManagement::LogEvent, priority section
+    reader.Init(aReader);
+
+    VerifyOrExit(envelope->mNumFieldsToRead > 0, err = CHIP_END_OF_TLV);
```

Comment:
nitpicky (saying this is reasonable enough is ok as well): Could there  be invalid events having repeated Tag_PriorityLevel for example and then a counter does not work as well?

Should we be using a bitflag here of 'field read' instead of a counter?

Reply:
Done, thanks, update with bitflag.